### DISCUSSION TITLE
Kernel Module 内存问题排查 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by SF-Zhou** [SF-Zhou/sf-zhou.github.io#155] [Announcements]
https://sf-zhou.github.io/linux/kernel_memory_usage_troubleshooting.html

**Comment by SF-Zhou** [comment]
如果你对存储、内核、C++ 感兴趣并且最近在看机会，我应该可以帮你内推。坐标杭州，行业量化，[点我发送简历](mailto:sf.zhou@high-flyer.cn)。同时在招 AI 平台开发和深度学习工程师，欢迎丢简历～

**Comment by zchrissirhcz** [comment]
有一个typo，泄露 => 泄漏。

问题的原因，我理解为: 原来每次记录inode都是新申请内存，现在一次性申请一大块内存，不知是否正确：

```c++
struct ListNode { ListNode* next; int val; };
const int g_amount = 10000;
for (int i = 0; i < g_amount; i++)
{
  ListNode* node = (ListNode*)malloc(sizeof(ListNode));
  node->val = inode;
}
```

现在改为了
```c++
const int g_amount = 10000;
int* inode_records = (int*)malloc(sizeof(int)*g_amount);
```

**Reply by SF-Zhou** [reply]
typo 已修改，谢谢～

由于是公司的源码，这里没有阐述地特别清晰。可以认为是有一个元信息列表，挂在 inode 结构里。修改是从原先的双向链表替换为数组，双向链表的节点类型为：
```c++
struct ListNode {
  void *data;
  ListNode *prev;
  ListNode *next;
};
```
这里三个指针会用 `kmalloc-32` 分配，但实际上需要存储的数据 `uint16_t` 只需要 2 个字节，这里会相差 16 倍。

**Reply by zchrissirhcz** [reply]
明白了。感谢分享！
